extension/ballastextension/memory_ballast.go

```diff
@@ -25,17 +25,32 @@ import (
 const megaBytes = 1024 * 1024
 
 type memoryBallast struct {
-	cfg     *Config
-	logger  *zap.Logger
-	ballast []byte
+	cfg       *Config
+	logger    *zap.Logger
+	ballast   []byte
+	memGetter func() (uint64, error)
 }
 
 func (m *memoryBallast) Start(_ context.Context, _ component.Host) error {
+	var ballastSizeBytes uint64
+	// absolute value supersedes percentage setting
 	if m.cfg.SizeMiB > 0 {
-		ballastSizeBytes := uint64(m.cfg.SizeMiB) * megaBytes
+		ballastSizeBytes = m.cfg.SizeMiB * megaBytes
+	} else {
+		totalMemory, err := m.memGetter()
+		if err != nil {
+			return err
+		}
+		ballastPercentage := m.cfg.SizeInPercentage
```

Comment:
I might miss it but is the config value validated?

Reply:
Added `config.Validate()`